tests/cpp/test_multidevice_transformer.cpp

```diff
@@ -984,6 +998,275 @@ TEST_P(DistributedTransformerTest, Forward) {
+  TensorView* ln0_w = makeContigTensor(1);
+  TensorView* ln0_b = makeContigTensor(1);
+  TensorView* ln0_mean = makeConcreteTensor({256, 1});
+  TensorView* ln0_rstd = makeConcreteTensor({256, 1});
+  TensorView* mha_linear1 = makeContigTensor(2);
+
+  fusion->addInput(x);
+  fusion->addInput(grad);
+  fusion->addInput(mha_w0);
+  fusion->addInput(mha_b0);
+  fusion->addInput(mha_w1);
+  fusion->addInput(mha_b1);
+  fusion->addInput(mlp_w0);
+  fusion->addInput(mlp_b0);
+  fusion->addInput(mlp_w1);
+  fusion->addInput(mlp_b1);
+  fusion->addInput(mlp_mask);
+  fusion->addInput(mha_mask);
+  fusion->addInput(mha_sdpa_out);
+  fusion->addInput(mha_sdpa_log_sumexp);
+  fusion->addInput(mha_sdpa_seed);
+  fusion->addInput(mha_sdpa_offset);
+  fusion->addInput(ln1_w);
+  fusion->addInput(ln1_b);
+  fusion->addInput(ln1_mean);
+  fusion->addInput(ln1_rstd);
+  fusion->addInput(ln0_w);
+  fusion->addInput(ln0_b);
+  fusion->addInput(ln0_mean);
+  fusion->addInput(ln0_rstd);
+  fusion->addInput(mha_linear1);
+
+  const auto D = mha_w0->axis(0)->extent()->value().as<int64_t>();
+  // Recomputation
+  auto ln_0 = layer_norm(x, norm_shape, ln0_w, ln0_b, eps);
+  auto mha_in = castOp(dtype, std::get<0>(ln_0));
+  auto qkv = mha_qkv(mha_in, mha_w0, mha_b0, mha_w1, mesh);
+  // The thunder trace recompute mha linear1, but this would result in 3
+  // AllReduces in the backwards pass.
+  if (mha_linear1 == nullptr) {
+    TensorView* sdpa_transpose = transpose(mha_sdpa_out, 2, 3);
+    TensorView* sdpa_reshape =
+        reshape(sdpa_transpose, {D, B, S, H / D, E / H}, {D, B * S, E / D});
+    TensorView* mha_local_matmul1 = matmul(sdpa_reshape, mha_w1);
+    TensorView* mha_matmul1 = sum(mha_local_matmul1, {0}); // allreduce
+    mha_linear1 = add(mha_matmul1, broadcast(mha_b1, {true, false}));
+  }
+  Val* scale = IrBuilder::create<Val>(1.0 / (1.0 - kDropoutProb));
+  // Use input mha mask to implement dropout
+  auto mha_out = mul(mha_linear1, mha_mask);
+  mha_out = mul(mha_out, scale);
+  auto resid_0 = add(x, mha_out);
+  auto ln_1 = layer_norm(resid_0, norm_shape, ln1_w, ln1_b, eps);
+  auto mlp_in = castOp(dtype, ln_1.output);
+  // Note: We only use linear0 and gelu outputs from the mlp forward pass.
+  auto mlp_tensors = mlp(mlp_in, mlp_w0, mlp_b0, mlp_w1, mlp_b1, mesh);
+
+  // Backwards
+  auto mlp_grads = mlp_backwards(
+      grad,
+      mlp_in,
+      mlp_mask,
+      mlp_w0,
+      mlp_b0,
+      mlp_w1,
+      mesh,
+      mlp_tensors[0],
+      mlp_tensors[1]);
+  auto ln1_grads = layer_norm_backward(
+      castOp(DataType::Float, mlp_grads[6]),
+      resid_0,
+      norm_shape,
+      ln1_mean,
+      ln1_rstd,
+      ln1_w,
+      ln1_b,
+      {true, true, true});
+  auto mha_grads = mha_backwards(
+      mha_in,
+      mha_w0,
+      mha_b0,
+      mha_w1,
+      mha_mask,
+      mha_sdpa_out,
+      mha_sdpa_log_sumexp,
+      mha_sdpa_seed,
+      mha_sdpa_offset,
+      ln1_grads.grad_input,
```

Comment:
Don't we need to add `grad` on top of this? In the forward pass, the result of MHA is forked to `ln1` and the `add` at the end. So in backprop I expect the two corresponding gradients to be added.

Reply:
let me check the thunder trace more closely, I might have missed that.

Reply:
Thanks! I was missing both the residual adds in the back prop.
Also the forward pass had the wrong tensors added in the residuals which is also fixed.